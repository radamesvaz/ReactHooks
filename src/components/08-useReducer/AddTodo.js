import React from 'react'
import { useForm } from '../../hooks/useForm'

export const AddTodo = ({ handleAddTodo }) => {

    const [ { desc }, handleInputChange, reset ] = useForm({
        desc: ''
    })

    const handleSubmit = (e) =>{
        e.preventDefault();

        if(desc.trim().length <= 1) return;

        const newTodo = {
            id: new Date().getTime(),
            desc: desc,
            done: false
        }

        handleAddTodo(newTodo);
        reset();
    }

    return (
        <form onSubmit={ handleSubmit }>
        <input 
            type="text"
            name="desc"
            placeholder="Agregar..."
            className="form-control"
            autoComplete="off"
            onChange={ handleInputChange }
            value={desc}
        />

        <button
            type="submit"
            className="btn btn-outline-primary mt-1 btn-block"
        >
            Agregar
        </button>

    </form>
    )
}
