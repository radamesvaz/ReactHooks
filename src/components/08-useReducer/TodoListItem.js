import React from 'react';
import PropTypes from 'prop-types';

export const TodoListItem = ({ todo, index, handleDeleteItem, handleToggle }) => {
    return (
        <li className="list-group-item">
            <p
                className={` ${todo.done && 'complete'} `}
                onClick={ () => handleToggle(todo.id) }
            >
               {index + 1}. {todo.desc}
            </p>

            <button
                className="btn btn-danger"
                onClick={ () => handleDeleteItem(todo.id) }
            >
                Eliminar
            </button>
        </li>
    )
}

TodoListItem.propTypes = {
    todo: PropTypes.object.isRequired,
    index: PropTypes.number.isRequired,
    handleDeleteItem: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired
}