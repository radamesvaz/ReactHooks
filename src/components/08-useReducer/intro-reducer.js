const initialState = [{
    id: 1,
    todo: 'Irse de venezuela',
    status: false
}],

 /*const*/ todosReducer = ( state = initialState, action ) => {

    if(action?.type) return [...state, action.payload]

    return state
}

const newTodo = [{
    id: 2,
    todo: 'Ganar plata',
    status: false
}];

const reducerAction = {
    type: 'agregar',
    payload: newTodo
}

let todo = todosReducer();

todo = todosReducer(initialState, reducerAction)



console.log(todo)