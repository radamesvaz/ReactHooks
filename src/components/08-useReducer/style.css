body{
    padding: 70px;
    background-color: black;
    color: #C724B1;
}

h1{
    color: #04d9ff ;
}

p{
    cursor: pointer;
    margin: 0px;
}

.list-group-item{
    align-items: center;
    background-color: black;
    display: flex;
    justify-content: space-between;
}

.complete{
    text-decoration: line-through;
}