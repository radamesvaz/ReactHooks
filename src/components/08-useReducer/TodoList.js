import React from 'react';
import PropTypes from 'prop-types';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({ todos, handleDeleteItem, handleToggle }) => {
    return (
        <ul>
            {
                todos.map( (todo, index) => (
                    <TodoListItem 
                        key={todo.id}
                        todo={todo}
                        index={index}
                        handleDeleteItem={handleDeleteItem}
                        handleToggle={handleToggle}
                    />
                ))
            }
        </ul>
    )
}

TodoList.propTypes = {
    todos: PropTypes.array.isRequired,
    handleDeleteItem: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired
}