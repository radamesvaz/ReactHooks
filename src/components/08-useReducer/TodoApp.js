import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';

import { TodoList } from './TodoList';
import { AddTodo } from './AddTodo';

import './style.css';


// Si no se consiguen items con el key value 'todos' se va a devolver un arreglo vacío como estado por defecto
const init = () => JSON.parse(localStorage.getItem('todos')) || [];

export const TodoApp = () => {

    const [ todos, dispatch ] = useReducer(todoReducer, [], init);

    // console.log(todos)
    // console.log(desc)

    //  Este useEffect se va a disparar cada vez que haya cualquier cambio en los todos, y los va a guardar en el sessionStorage
    useEffect( () => {
        localStorage.setItem('todos', JSON.stringify(todos))
    }, [todos])

    const handleAddTodo = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        })
    }

    const handleDeleteItem = ( todoId ) => {
        //  Crear la acción
        const action = {
            type: 'delete',
            payload: todoId
        }

        //  dispatch    
        dispatch(action);
    }

    const handleToggle = ( todoId ) => {
        //  el dispatch también puede enviarse asi:
        dispatch({
            type: 'toggle',
            payload: todoId
        })
    }

    return (
        <div>
            <h1>TodoApp: { todos.length } </h1>
            <hr />

        <div className="row">

            <div className="col-7">
                Todos
                {/* Aquí va ir un componente TodoList, se le pasan los todos (estado) handleDelete handleToggle */}
            <TodoList 
                todos={todos} 
                handleDeleteItem={handleDeleteItem} 
                handleToggle={handleToggle} 
            />

            </div>

            <div className="col-5">
                <h4>Agregar Todo</h4>
                <hr />
                <AddTodo 
                    handleAddTodo={ handleAddTodo } 
                />
            </div>

        </div>

        </div>
    )
}

/* 
    Hay mucho que decir aquí, esta lección me costo. Pero el punto principal de la clase es el useReducer: Recibe la función reducer como primer argumento
    el estado inicial como segundo argumento, y un tercer argumento que es opcional, y es la inicialización de una función que va a retornar el estado inicial del reducer.

    RECORDAR: cada vez que hay un cambio en el estado, react vuelve a ejecutar lo que se encuentra arriba del return a ver si hay cambios importantes. Quitar comentarios de los console.log
    para observar mejor en líneas 15 y 16. Repasar el uso del useForm

    Como el sessionStorage solamente acepta strings, hay que aplicarle el json.stringify(). Y para recuperar el objeto en su estado normal, se usa el json.parse()

    VIDEO: 128
*/
