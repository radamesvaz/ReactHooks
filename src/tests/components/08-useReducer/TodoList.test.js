import { shallow } from 'enzyme';
import React from 'react';
import { TodoList } from '../../../components/08-useReducer/TodoList';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Haciendo pruebas sobre el componente <TodoList />', () => {
    
    const handleDeleteItem = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow(
        <TodoList 
            todos={ demoTodos }
            handleDeleteItem={ handleDeleteItem }
            handleToggle={ handleToggle }
        />
    )


    test('debe mostrar el componente correctamente ', () => {
        
        expect(wrapper).toMatchSnapshot();
    });

    test('debe de tener dos <TodoListItem /> ', () => {
        
        expect( wrapper.find('TodoListItem').length ).toBe( demoTodos.length );

        expect( wrapper.find('TodoListItem').at(0).prop('handleDeleteItem') ).toEqual( expect.any(Function) )
    })
    
    
})
