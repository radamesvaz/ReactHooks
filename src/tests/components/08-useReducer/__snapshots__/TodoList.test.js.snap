// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Haciendo pruebas sobre el componente <TodoList /> debe mostrar el componente correctamente  1`] = `
<ul>
  <TodoListItem
    handleDeleteItem={[MockFunction]}
    handleToggle={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "desc": "hacer ejercicios",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleDeleteItem={[MockFunction]}
    handleToggle={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "desc": "tener cuadritos",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
